### ISSUE TITLE
JAX fitting paths fill masked cells with mean resistance while the forward path fills them with a barrier

### ISSUE THREAD

**Issue by seansorek** [seansorek/DiffiScape#134]
## Description

Every JAX **fitting** path fills masked-out (invalid) grid cells with the **mean resistance of the valid cells**, while the JAX **forward/evaluation** path fills the same cells with a **barrier**. The model is therefore fit on a landscape whose no-data cells conduct at average permeability, and then evaluated on a landscape where those same cells are effectively impassable.

Forward path — `ds_jax_connectivity()` turns no-data into a barrier before handing the surface to Python:

```r
# R/jax_bridge.R:192-199
# nodata -> barrier (minimum conductance).
if (parameterization == "permeability") {
  R_mat[is.na(R_mat)] <- 0        # clamped up to p_min = 1/5000
} else {
  R_mat[is.na(R_mat)] <- 1e9      # clamped down to r_max = 5000
}
```

Fitting paths — all three embed valid-cell resistance into the full grid over a **mean** background:

```python
# inst/python/diffiscape_jax/core.py:621-623  (_connectivity_objective, parametric)
full_surface = jnp.ones(n_rows * n_cols) * jnp.mean(resistance_flat)
full_surface = full_surface.at[valid_mask].set(resistance_flat)
surface_2d = full_surface.reshape((n_rows, n_cols))
```

```python
# inst/python/diffiscape_jax/optimize.py:490-493  (run_neural_optimization loss_fn)
fill_val = jnp.mean(resistance)
full_surface = jnp.ones(n_rows * n_cols) * fill_val
full_surface = full_surface.at[mask_jnp].set(resistance)
```

```python
# inst/python/diffiscape_jax/sample.py:110-113  (_build_numpyro_model, NUTS + ADVI)
fill_val = jnp.mean(resistance)
full_surface = jnp.ones(n_rows * n_cols) * fill_val
full_surface = full_surface.at[valid_mask].set(resistance)
```

The mask is real, not hypothetical: `valid_mask <- stats::complete.cases(basis_matrix)` (`R/optimizer.R:257`, `R/optimizer.R:394`), so every cell with an `NA` in any basis layer — water, area outside the study boundary, a covariate with a smaller footprint — is masked. `inst/python/diffiscape_jax/tests/test_neural_optimize.py:362` exercises exactly this case with `valid[[0, 5, 17]] = False`.

`model_type = "conv"` has the same problem one level up: `.optimize_neural()` fills invalid cells of the basis grid with `colMeans(basis_values)` (`R/optimizer.R:423`), so the conv/IRL full-grid forward pass predicts an ordinary mid-range resistance for cells that are barriers at evaluation time.

**Location:**
- [R/jax_bridge.R#L192-L199](https://github.com/seansorek/DiffiScape/blob/252089e9a8b73288d4e151734c53cc67bcb8b00d/R/jax_bridge.R#L192-L199)
- [inst/python/diffiscape_jax/core.py#L621-L623](https://github.com/seansorek/DiffiScape/blob/252089e9a8b73288d4e151734c53cc67bcb8b00d/inst/python/diffiscape_jax/core.py#L621-L623)
- [inst/python/diffiscape_jax/optimize.py#L490-L493](https://github.com/seansorek/DiffiScape/blob/252089e9a8b73288d4e151734c53cc67bcb8b00d/inst/python/diffiscape_jax/optimize.py#L490-L493)
- [inst/python/diffiscape_jax/sample.py#L110-L113](https://github.com/seansorek/DiffiScape/blob/252089e9a8b73288d4e151734c53cc67bcb8b00d/inst/python/diffiscape_jax/sample.py#L110-L113)

## Impact

This is the same class of defect as #105 / #123 — the objective optimises a different connectivity definition than the one every downstream step evaluates — but on the *mask* axis rather than the source-topology axis.

In a moving-window Omniscape solve, a masked cell inside a window's buffer is not a cosmetic detail: it determines whether current can route through that part of the window at all. Filling it at mean permeability opens a channel that is closed at evaluation time. Concretely:

- The fitted basis coefficients `z` and the intercept `r_0` are the argmax of a likelihood computed on the wrong landscape, so the returned resistance surface is not the MLE of the model the package then reports.
- `gamma` (the connectivity coefficient in `log lambda = alpha + gamma * log(1 + C)`) is calibrated against a `C` whose scale and spatial pattern differ from the `C` produced by `ds_jax_connectivity()` / `evaluate_full_model()`.
- The NUTS/ADVI posterior in `sample.py` inherits the same misspecification, so credible intervals describe a landscape that is never evaluated.
- Severity scales with how much of the extent is masked. On a study area with a large water body or a ragged boundary — routine in landscape genetics — a large fraction of the grid is affected.

The gradient/sampling paths do not fail loudly; they converge to a plausible-looking surface, so this is silent.

## Suggested fix

Make the fill a single shared function so the two paths cannot drift again, mirroring how #105 was resolved by sharing `cumulative_current_core` between the objective and the forward path.

The forward path's convention is the correct one (no-data is a barrier), so the fitting paths should match it: fill invalid cells with `r_max` in resistance mode and `p_min` in permeability mode rather than `jnp.mean(...)`. Note that a constant fill also removes the current (unintended) dependence of the background on the parameters — `jnp.mean(resistance_flat)` is differentiable, so today a gradient step on `z` moves the background resistance of every masked cell as a side effect.

Two follow-ons worth deciding at the same time:

1. Do the same for the conv/IRL basis fill in `.optimize_neural()` (`colMeans(basis_values)`), or accept the full-grid prediction there but overwrite the masked cells with the barrier value before the permeability conversion.
2. Add a regression test in the spirit of `test_objective_and_forward_path_agree_on_connectivity_direction` (`inst/python/diffiscape_jax/tests/test_core.py:346`) that runs a grid with some invalid cells through both the objective and `window.cumulative_current()` and asserts the two connectivity surfaces agree — the current tests either use all-valid fixtures or only check the *shape* of the returned array.

---
*Auto-detected by codebase-review agent*

**Comment by seansorek** [comment]
**Triage — confirmed scientific correctness bug, Critical severity**

Bug is real and well-documented. The issue cites specific code lines in `core.py`, `optimize.py`, `sample.py`, and `R/optimizer.R` that all fill masked (invalid) grid cells with `jnp.mean(resistance_flat)` — a value that changes with the fitted parameters — while the forward/evaluation path (`ds_jax_connectivity()` via `R/jax_bridge.R:192-199`) fills the same cells with a barrier (`0` permeability / `1e9` resistance).

**Confirmed impact:**

The three fitting paths (parametric gradient, neural, NUTS/ADVI) converge on parameter estimates that are optimal for a landscape that no downstream step ever evaluates. `gamma` (the connectivity coefficient) is calibrated against a `C` that differs in both scale and spatial structure from what `evaluate_full_model()` computes. On any study area with water bodies, a ragged boundary, or covariate gaps — the routine case in landscape genetics — a substantial fraction of the grid is masked, so the mismatch is proportionally large.

Additionally: the `jnp.mean(resistance_flat)` fill is differentiable, meaning gradient steps on the basis coefficients `z` also move the background resistance of every masked cell as an unintended side effect. This is a gradient-corruption issue on top of the model-mismatch issue.

**Severity:** Critical — this is the same class of defect as #105/#123 (objective optimizes a different connectivity surface than the one evaluated downstream), but on the mask axis. The result is biased parameter estimates that appear plausible, with no error or warning. The NUTS/ADVI credible intervals describe a model that is never computed. Affects every real-world use case with masked cells.

**Fix scope:** Moderate. The suggested fix (replace `jnp.mean(...)` with a barrier fill in all four locations, extracted into a shared helper) is straightforward but touches 4 files across R and Python. A regression test pairing the objective's internal connectivity against `window.cumulative_current()` on a grid with masked cells is the key guard to add (mirroring `test_objective_and_forward_path_agree_on_connectivity_direction` in `test_core.py`). The R side also needs to handle the conv/IRL basis fill in `.optimize_neural()`.

No open PR currently addresses this.

**Comment by seansorek** [comment]
Fixed by #135.
